[충남대 FE 김은수] 3단계 단체 주문 기능 구현하기

shushuburger opened the pull request:
충남대학교 FE 12팀 김은수입니다.

### 요구사항
- [x] 본인만의 기준으로 일관된 코드를 작성해주세요.
- [x] 기능 단위로 나누어 커밋을 해주세요.
- [x] 재사용 가능성이 높은 컴포넌트, Hooks가 보이면 재사용성을 고려하여 만들어요.
- [x] 주문하기 페이지의 폼을 React Hook Form를 사용해서 리팩터링해요. (form 데이터를 useState로 다루는 경우는 없게 해요.)
- [x] 받는 사람 기능 고도화
  - [x] [참고 URL](https://kakaotech-mission2-order.pages.dev/)을 살펴보고 받는 사람 기능의 UI를 구현해요.
  - [x] 받는 사람을 최대 10명까지 등록 가능하게 해요.
  - [x] 10명의 정보가 정확하게 입력되지 않으면 정보가 반영되지 않아요.
  - [x] 받는 사람의 전화번호가 중복되면 안돼요.
  - [x] 최소 선택 수량은 1개 이상이어야 해요.
  - [x] 전화번호는 01012341234 형태로만 받을 수 있어요.
- [x] 주문하기 페이지 고도화
  - [x] [참고 URL](https://kakaotech-mission2-order.pages.dev/)을 살펴보고 변경된 기능을 구현해요.

-------------------------------------------------
- 상수 처리, 중복 코드 제거, rendering 부분 단순화, 컴포넌트 분리, 타입 단언 주의 등에 집중하며 작업해보려고 노력했습니다.

shushuburger committed:
docs(readme): 3단계 때 구현해야 하는 내용 입력

shushuburger committed:
feat(receiver): 받는 사람 폼 입력 UI 기본 작업

- ReceiverForm.tsx에 기본적인 UI 구조를 작업해주었습니다.
- 추가 버튼, 받는 사람 입력하는 등 기본적인 동작을 구현해주었습니다.

shushuburger committed:
feat(order): 받는 사람 입력 모달 기능 구현 및 개선

- 받는 사람 정보를 입력하는 모달 UI 및 상태 로직을 구현해주었습니다.
- '추가하기' 클릭 시 이름/전화번호/수량 필드 포함한 입력 항목이 추가되도록 했습니다.
- 최대 10명까지 제한하면서 각 항목을 삭제할 수 있도록 했습니다.
- 입력 필드가 화면을 초과할 경우 스크롤을 적용하여 튀어나가지 않도록 했습니다.
- "N명 완료" 버튼이 현재 입력 수에 따라 갱신되도록 해주었습니다.
- README에 구현 완료한 내용을 표시해주었습니다.

shushuburger committed:
feat(order): 받는 사람 입력 모달 기능 개선

- '취소' 시 모달 입력 상태 초기화 처리

shushuburger committed:
feat(order): 받는 사람 정보 완료 후 수정 가능하도록 상태 관리 추가

- 'n명 완료' 클릭 시 입력값을 유지하면서 모달만 닫히도록 변경해주었습니다.
- 완료 이후 모달 트리거 버튼 텍스트를 '추가'에서 '수정'으로 변경해주었습니다.
- 설정한 받는 사람의 모든 정보가 정상적으로 입력되어야지만 완료 버튼을 통해 모달을 닫을 수 있도록 했습니다.
- 받는 사람의 전화번호 중복 여부를 걸러낼 수 있도록 했습니다.
- 수량 조건, 전화번호 조건 등 기존의 검사 로직을 이용해 구현했습니다.
- README에 구현 완료 내용을 표시했습니다.

shushuburger committed:
refactor(order): 받는 사람 정보 구조 및 UI 개선

- useOrderForm 훅에서 receiver 관련 내용을 제거해주어 정상적으로 검사가 가능하도록 했습니다.
- FormValues를 senderName, textMessage만 포함하도록 해주었습니다.
- validateForm이 sender/message만 검사하도록 해주었습니다.
- ReceiverForm에서 완료 후 받는 사람 정보를 테이블 형태로 출력되도록 UI를 개선했습니다.
- README에 구현 완료한 내용들을 표시했습니다.

shushuburger committed:
refactor(order): 주문하기 페이지 폼을 React Hook Form 기반으로 전면 리팩토링

- useForm, useFieldArray를 사용해 senderName, textMessage, receivers 상태를 RHF로 통합해주었습니다.
- useState 기반 폼 상태 및 수동 유효성 검사 함수를 제거했습니다.
- 모든 입력 필드를 forwardRef 구조로 변경하여 register()와 연결 가능하게 구성해주었습니다.
- SenderForm, MessageInput, InputField 컴포넌트를 RHF와 호환되도록 구조 개선해주었습니다.
- ReceiverForm은 useFieldArray로 배열 상태와 유효성 검사를 통합해주었습니다.
- 전화번호 중복 검사를 register의 validate로 옮겨서 실시간 검증을 하도록 했습니다.
- 주문 제출 시 handleSubmit을 통해 RHF 방식으로 유효성 검증 및 데이터를 수집하도록 했습니다.
- README에 구현 완료한 내용들을 표시했습니다.

shushuburger committed:
refactor(receiver): ReceiverFormSection으로 컴포넌트들 분리

- ReceiverForm을 5개의 파일로 분리하여 책임 분리 및 재사용성을 향상해보려 했습니다.
  - useReceiverForm.ts: RHF 로직 및 중복 검사 분리
  - ReceiverInputItem.tsx: 개별 입력 필드 컴포넌트 추출
  - ReceiverModal.tsx: 모달 UI 및 입력 리스트 분리
  - ReceiverTable.tsx: 수신자 리스트 테이블 컴포넌트 추출
  - ReceiverForm.tsx: 위 컴포넌트 조합 및 상태 관리 전담
- README에 구현 완료 내용을 표시했습니다.

shushuburger committed:
refactor(receiver-form): 상수화, 토큰 적용, 렌더링 정리 등 구조 개선

- 하드코딩된 문자열 및 숫자 값을 상수로 추출하여 가독성과 유지보수성을 향상시켰습니다.
- 색상 및 여백 값에 디자인 토큰(theme.color, theme.spacing 등) 적용했습니다.
- 중복 렌더링 구조를 map 반복문으로 통합하여 반복 코드를 최대한 제거했습니다.
- README에 구현 완료를 표시했습니다.

shushuburger committed:
fix(receiver-form): 수정 버튼 클릭 시 받는 사람 입력값이 초기화되는 문제 수정

- ReceiverForm에서 ReceiverModal에 initialValues={receiverList} 전달하도록 수정했습니다.
- useReceiverForm 내부 useEffect에서 reset을 통해 기존 수신자 목록을 입력 필드에 초기값으로 세팅해주도록 했습니다.
- 수정하기 버튼 클릭 시 이전에 입력한 내용이 유지되도록 개선하였습니다.

shushuburger committed:
style(receiver-form): 수정 버튼을 회색 배경과 검은 글씨로 스타일 변경

- AddButton 컴포넌트에 token을 활용해 텍스트 색상을 적용해주었습니다.

Leejin-Yang commented on src/components/ReceiverFormSection/ReceiverModal.tsx:
```diff
@@ -0,0 +1,187 @@
+  initialValues: Receiver[];
+}
+
+const ReceiverModal = ({ onConfirmList, onClose, initialValues }: Props) => {
```
모달의 경우 portal을 사용해 부모에 영향을 받지 않도록 루트 밖으로 빼서 작성할 수 있어요. 아래 링크 한 번 참고해보세요 😊

https://react.dev/reference/react-dom/createPortal

shushuburger replied:
작성해주신 문서 참고해 작업해보았습니다!
공식 문서에서는 createPortal을 사용하고 한 파일 내에서 렌더링 코드까지 다루고 있었지만 모달이라는 형태를 반복해서 사용할 수도 있겠다는 생각이 들어 아예 ModalPortal.tsx로 분리하여 작성해주었습니다.

shushuburger replied:
```
  const onSubmit = (data: FormValues) => {
    if (receiverList.length === 0) {
      alert('받는 사람을 한 명 이상 추가해주세요.');
      return;
    }

    alert(
      `주문이 완료되었습니다.\n` +
        `상품명: ${product.name}\n` +
        `구매 수량: ${totalQuantity}\n` +
        `발신자 이름: ${data.senderName}\n` +
        `메시지: ${data.textMessage}`
    );
    navigate(ROUTES.HOME);
  };
```

다음과 같이 받는 사람이 0명인지 아닌지를 판단할 수 있도록 작업해주었습니다!

Leejin-Yang commented on src/pages/OrderPage.tsx:
```diff
@@ -1,118 +1,95 @@
-    const isValid = validateForm();
-    if (!isValid) return;
-
+  const onSubmit = (data: FormValues) => {
```
현재 받는 사람이 없어도 주문을 할 수 있네요. 이 부분 한번 확인해주세요.!

<img width="453" height="229" alt="Screenshot 2025-07-11 at 18 07 48" src="https://github.com/user-attachments/assets/286b41d7-3f32-4ed6-a156-999eb2296bd3" />

shushuburger replied:
참고 링크에서 받는 사람 없이도 주문이 성공적으로 동작하는 것을 보고
당시에는 '과제 링크의 기능 로직을 따라야 한다!'라는 생각이 더 강해서 이 부분을 크게 신경쓰지 않았습니다.
금방 수정해보겠습니다!!

Leejin-Yang commented on src/pages/OrderPage.tsx:
```diff
@@ -1,118 +1,95 @@
-    validateForm,
-    totalPrice,
-  } = useOrderForm(product.price.sellingPrice);
+    register,
+    handleSubmit,
+    setValue,
+    formState: { errors },
+  } = useForm<FormValues>({
+    defaultValues: {
+      senderName: '',
+      textMessage: messageCards[0].defaultTextMessage,
+    },
+  });
+
+  const [receiverList, setReceiverList] = useState<
+    { name: string; phone: string; quantity: number }[]
+  >([]);
```
이 두개를 함께 관리한다면 submit 과정에서 받는 사람 유효성도 검사할거라 생각이 드네요

shushuburger replied:
현재 "받는 사람이 0명일 경우 주문 완료를 방지한다!"라는 것까지 구현하고 나서의 상황입니다!

ReceiverForm.tsx과 ReceiverModal.tsx
- 받는 사람의 이름, 전화번호 중복 유무, 수량 등에 대한 유효성을 점검
- 모든 유효성이 통과되어야지만 성공적으로 반영됨
- 받는 사람이 0명일 경우는 신경쓰지 않음
- 작성을 하다가 "취소"를 하는 경우도 있기 때문에 여기서 "0명"일 경우를 유효하지 않다고 받아들지 않아야 한다고 생각함

OrderPage.tsx
- 보내는 사람, 축하메시지, **받는 사람이 0명인 경우**만 유효성 검사
- 이미 "받는 사람이 1명 이상일 경우" 그 정보는 유효성 검사를 통과한 정보들이 주어짐

다음과 같은 상황이라서 저는 "유효성 검사를 다시 할 필요가 없는 요소들에 대해서는 한 번 더 검사를 진행하는게 비효율적인 것 같다"라고 받아들였습니다!

shushuburger replied:
rhf에 대해서 더 알아보니 묶어서 관리해주는 것이 반드시 "유효성 검사를 한 번 더 한다"가 아니라고 하는 것을 새로이 알게되었습니다!!

FormProvider를 통한 prop drilling 문제를 해결하면서 OrderPage.tsx에서 모달의 내용까지 묶어서 점검하도록 해보겠습니다!!

Leejin-Yang commented on src/pages/OrderPage.tsx:
```diff
@@ -1,118 +1,95 @@
+            {...register('textMessage', {
+              required: ERROR_MESSAGES.EMPTY_MESSAGE,
+            })}
+            error={errors.textMessage?.message}
           />
           <SenderForm
-            name="senderName"
-            value={String(formValues.senderName)}
-            onChange={handleInputChange}
-            error={formErrors.senderName}
+            {...register('senderName', {
+              required: ERROR_MESSAGES.EMPTY_SENDER,
+            })}
+            error={errors.senderName?.message}
```
FormProvider를 사용해 컨텍스트를 사용해 props 뎁스를 줄여볼 수 있어보여요 😊

shushuburger replied:
ReceiverForm을 먼저 작업한 후 SenderForm, MessageInput에도 FormProvider 양식을 적용해주었습니다.

Leejin-Yang commented on src/components/OrderSection/SenderForm.tsx:
```diff
@@ -1,29 +1,27 @@
-    </Wrapper>
-  );
-};
+const SenderForm = forwardRef<HTMLInputElement, SenderFormProps>(
```
컨텍스트를 사용하면 이 단계에서 메서드들을 가져올 수 있어보여요 😊

Leejin-Yang commented on src/components/ReceiverFormSection/ReceiverInputItem.tsx:
```diff
@@ -0,0 +1,113 @@
+  errors,
+  isDuplicate,
+}: Props) => {
+  const fields = [
```
필드를 잘 구성해주셨네요 👍
이 필드를 생성하는 함수를 컴포넌트 외부에서 선언해 사용해보는건 어떨까요?

shushuburger replied:
필드 관련 코드를 외부로 옮겨주었습니다!

Leejin-Yang reviewed (changes requested):
은수님 안녕하세요 👋
이번 미션 마지막 단계까지 잘 오셨네요 👍
전체적으로 중복을 줄이려고 노력하신게 보여서 좋았습니다 👍

이번 단계 진행하면서 useState를 라이브러리로 리팩터링한 과정이 어땠는지 은수님 이야기를 듣고 싶은데, 코멘트 남겨주시면 감사하겠습니다 😊

그리고 받는 사람이 없을 때 주문이 성공하는 걸로 보여서 이 부분 우선적으로 확인해주세요
코멘트 몇개 남겼는데 확인해주시구 리뷰 요청 보내주세요 🙌

shushuburger committed:
fix(order): 받는 사람이 없을 경우 주문 방지

- receiverList가 비어 있으면 주문 제출이 되지 않도록 handleSubmit 내부에 검사 로직을 추가했습니다.
- 주문 완료 시 상품명, 총 수량, 발신자 이름, 메시지를 포함한 alert 메시지를 출력하도록 했습니다.

shushuburger committed:
refactor(modal): ReceiverModal에 React Portal 적용 및 ModalPortal 컴포넌트 생성

- ModalPortal 컴포넌트 생성하여 부모 div에 영향을 받지 않도록 렌더링되게 코드 작업을 했습니다.

shushuburger committed:
refactor(form): OrderPage에서 RHF 컨텍스트를 전역으로 관리하도록 구조 개선

- FormProvider를 이용해 prop drilling 문제를 해결하였습니다.
- 모달 내 수신자 입력 필드 상태도 전역 폼에 포함되어 하나의 form으로 관리하도록 했습니다.
- 유효성 검사 및 값 동기화를 OrderPage 폼 컨텍스트에서 일괄 수행 가능하도록 했습니다.

shushuburger committed:
refactor(order): SenderForm에 FormProvider 기반 RHF context 적용

- useFormContext를 사용해 senderName 필드 등록 및 에러 메시지를 처리하도록 했습니다.
- OrderPage에서 불필요한 register, error props 제거해주었습니다.

shushuburger committed:
refactor(receiver): 필드 정의 로직을 외부 함수로 분리하여 컴포넌트 간결화

- ReceiverInputItem 내부의 필드 구성 로직을 getReceiverFields 함수로 분리하여주었습니다.

shushuburger commented:
### useState를 라이브러리로 리팩터링한 과정

기존에는 "보내는 사람, 메시지, 받는 사람 목록" 등의 요소에 변경이 일어날 때마다 각각을 useState를 통해 기본 상태 설정, 변경 시 값 반영 등을 작업해주었습니다.
그것들을 묶어서 FormValue와 같이 통합하는 type을 만든다고 해도 요소가 증가할 때마다 코드가 길어지고 가독성이 떨어진다고 느껴지는 건 어쩔 수 없었던 것 같습니다.

그러던 중에 npm install을 통해 react hook form을 설치해준 후에 본격적인 리팩토링에 들어갔습니다.
이를 통해 제가 만들어두었던 useOrderForm.ts 같은 custom hook을 만들어서 상태 관리를 위한 작업을 해주었다면 rhf 도입을 통해 기존 내장 메서드를 사용할 수 있다는 점이 굉장히 신기했습니다.

그냥 단순히 hook을 이용해 나에게 조금 더 최적화해보자라는 취지로 custom hook을 쓰면서도 신기하기도 하고 코드가 읽기 쉬워진 것 같다라는 느낌이 있었는데 그마저도 rhf로 대체할 수 있다라는 점이 굉장히 저에게는 놀라우면서도 학습할 새로운 내용이 생겨 신경쓸 부분이었습니다.

온전히 제가 rhf를 이해하고 작업하기에는 아직 실력이 부족해서 이번 단계는 ai에 대한 의존도가 높았던 것 같습니다.
ai에게 코드 변경 방식을 물어보고 왜 이렇게 변경했는지, 어떤 점이 변경되었는지 (기존에 쓰던 코드에서 안 쓰는 부분이 생겼다면 왜 안 쓰는지), 이렇게 바꾸면 뭐가 더 좋은 점인지, 나의 경우에는 이러한 요소들이 있고 기존에 작업하는 동작은 이런 방식인데 그렇다면 어디에 rhf를 적용할 수 있을지 등을 되물어가는 식으로 학습을 진행했습니다.

Leejin-Yang commented on src/pages/OrderPage.tsx:
```diff
@@ -1,122 +1,89 @@
+      senderName: '',
+      textMessage: messageCards[0].defaultTextMessage,
+      receivers: [],
+    },
+  });
 
-  const [selectedCardId, setSelectedCardId] = useState(messageCards[0].id);
-  const selectedCard = messageCards.find(card => card.id === selectedCardId)!;
+  const { handleSubmit, setValue, watch } = methods;
 
-  useEffect(() => {
-    handleChange('textMessage', selectedCard.defaultTextMessage);
-  }, [selectedCardId]);
+  const receiverList = watch('receivers');
+  const totalQuantity = receiverList.reduce((sum, r) => sum + r.quantity, 0);
+  const totalPrice = product ? totalQuantity * product.price.sellingPrice : 0;
```
이 부분을 useOrderForm으로 분리할 수 있어보여요 😊

Leejin-Yang commented on src/components/ReceiverFormSection/ReceiverForm.tsx:
```diff
@@ -0,0 +1,87 @@
+
+  const receiverList = watch('receivers');
+  const [isModalOpen, setIsModalOpen] = useState(false);
+  const [isConfirmed, setIsConfirmed] = useState(receiverList.length > 0);
```
이 값은 receiverList의 값에 따라 파생되는 상태로 관리할 수 있어보여요

Leejin-Yang commented on src/components/ReceiverFormSection/ReceiverForm.tsx:
```diff
@@ -0,0 +1,87 @@
+  const { control, watch } = useFormContext<{ receivers: Receiver[] }>();
+  const { replace } = useFieldArray({ name: 'receivers', control });
+
+  const receiverList = watch('receivers');
```
useWatch 훅과 watch의 차이점도 한번 살펴보시면 좋아보여요 😊

https://react-hook-form.com/docs/usewatch

Leejin-Yang commented on src/components/ReceiverFormSection/ReceiverInputItem.tsx:
```diff
@@ -0,0 +1,131 @@
+    control,
+  } = useFormContext<{ receivers: Receiver[] }>();
+
+  const receivers = useWatch({ name: 'receivers', control });
```
여기서는 useWatch를 쓰셨네요. watch와 어떤 차이점이 있을까요?

Leejin-Yang commented on src/components/ReceiverFormSection/ReceiverModal.tsx:
```diff
@@ -0,0 +1,204 @@
+}
+
+const ReceiverModal = ({ initialValues, onConfirmList, onClose }: Props) => {
+  const methods = useForm<{ receivers: Receiver[] }>({
+    defaultValues: {
+      receivers: initialValues,
+    },
+    mode: 'onChange',
+    reValidateMode: 'onChange',
+  });
+
+  const { control, getValues, trigger, reset } = methods;
+
+  const { fields, append, remove } = useFieldArray({
+    name: 'receivers',
+    control,
+  });
```
이 친구 관련해서는 기존에 작성하신 useReceiverForm을 활용할 수 있지 않을까요?

Leejin-Yang commented on src/components/common/ModalPortal.tsx:
```diff
@@ -0,0 +1,21 @@
+  children: ReactNode;
+}
+
+const ModalPortal = ({ children }: ModalPortalProps) => {
+  const [modalRoot, setModalRoot] = useState<HTMLElement | null>(null);
+
+  useEffect(() => {
+    setModalRoot(document.body);
+  }, []);
+
+  if (!modalRoot) return null;
+
+  return createPortal(children, modalRoot);
+};
```
포탈 컴포넌트를 잘 작성해주셨네요 👍
더 나아가서, body가 아닌 별도의 root div(예를 들어 id='modal-root')를 생성해 모달 전용 루트를 만들어 계층을 나누어주는 방법도 생각해 볼 수 있습니다 😊

Leejin-Yang commented on src/components/ReceiverFormSection/ReceiverForm.tsx:
```diff
@@ -0,0 +1,87 @@
+  const { replace } = useFieldArray({ name: 'receivers', control });
+
+  const receiverList = watch('receivers');
+  const [isModalOpen, setIsModalOpen] = useState(false);
```
이건 추가적으로 학습할 수 있는 부분인데, 앞으로 여러 모달이 생길 때 이런 모달 관련 상태들을 계속 선언하게 될거에요. 중복되는 작업을 줄이기 위해 모달을 관리하는 훅을 만들어 보는 것도 좋은 방법이에요 😊

Leejin-Yang reviewed (approved):
이번 미션 잘 마무리 하신 걸로 보이네요 👍
앱을 실행했을 때 요구사항에 맞게 동작해서 이번 미션은 이만 마무리하겠습니다.!
폼 관련해서 몇가지 코멘트 남겼는데, 다음 미션 시작 전에 한번 고민해보면 좋아보여요 😊

직접 폼 관리를 해봄과 라이브러리를 활용하면서 많은 차이점을 느꼈을거라 생각해요. 두 과정 모두 쉽지 않은 과정이였을건데, 남겨주신 코멘트를 보니 이번 미션 요구사항을 잘 따라오신 것으로 보입니다.
ai 활용해 학습 방향도 잘 잡으신 것 같구요 👍 

앞으로도 라이브러리를 보며 어떤 불편한 점을 개선하기 위해 이 라이브러리가 나오게 되었는지, 제공해주는 메서드들은 어떻게 동작하는지 살펴보며 학습해 나가면 좋아보입니다 (구현 코드를 확인해보면 더 좋구요 😊) 그리고 라이브러리 도입에 대해서도 필요성과 러닝 커브를 고려해서 결정해보면 좋아보여요

이번 미션 고생 많으셨고, 앞으로도 미션 진행하면서 어려운 점이나 궁금한 점이 있다면 언제든지 DM 남겨주세요. 수고 많았어요 🙌
